data_pipeline/generic_web_api/helper.py

```diff
@@ -0,0 +1,181 @@
+
+
+# pylint: disable=too-many-instance-attributes,too-many-arguments
+class DynamicURLManager:
+    def __init__(
+            self,
+            url_excluding_configurable_parameters: str,
+            from_date_param: str = None,
+            to_date_param: str = None,
+            date_format: str = None,
+            next_page_cursor: str = None,
+            page_number_param: str = None,
+            offset_param: str = None,
+            page_size_param: str = None,
+            page_size: int = None,
+            compose_able_url_key_val: dict = None,
+            **kwargs
+    ):
+        self.url_excluding_configurable_parameters = (
+            url_excluding_configurable_parameters
+        )
+        self.from_date_param = from_date_param
+        self.to_date_param = to_date_param
+        self.date_format = date_format
+        self.next_page_cursor = next_page_cursor
+        self.page_number_param = page_number_param
+        self.offset_param = offset_param
+        self.page_size_param = page_size_param
+        self.page_size = page_size
+        self.compose_able_url_key_val = compose_able_url_key_val
+        self.type_specific_params = kwargs
+
+    def _get_url_separator(self):
+        url = self.url_excluding_configurable_parameters
+        if "?" in url:
+            if url.strip().endswith("&") or url.strip().endswith("?"):
+                url_separator = ""
+            else:
+                url_separator = "&"
+        else:
+            url_separator = "?"
+        return url_separator
+
+    def compose_url(self, parameters_key_value: dict):
+        url = self.url_excluding_configurable_parameters
+        url_separator = self._get_url_separator()
+        params = parse.urlencode(
+            {
+                key: value
+                for key, value in parameters_key_value.items() if key and value
+            }
+        )
+
+        return url + url_separator + params
+
+    def get_url(
+            self,
+            url_compose_param: UrlComposeParam,
+    ):
+        start_date = datetime_to_string(
+            url_compose_param.from_date, self.date_format
+        )
+        end_date = datetime_to_string(
+            url_compose_param.to_date, self.date_format
+        )
+        param_dict = dict((key, value) for key, value in [
+            (self.from_date_param, start_date),
+            (self.next_page_cursor, url_compose_param.cursor),
+            (self.to_date_param, end_date),
+            (self.page_number_param, url_compose_param.page_number),
+            (self.offset_param, url_compose_param.page_offset),
+            (
+                self.page_size_param,
+                url_compose_param.page_size or self.page_size
+            )
+            ] if key and value)
+        param_dict = {
+            **param_dict,
+            **self.compose_able_url_key_val
+        }
+
+        return self.compose_url(param_dict)
+
+
+def get_url_manager(url_source_type: str = None):
```

Comment:
would suggest to rename to `get_url_manager_class`, `_factory` or similar.

Maybe `url manager` could also be renamed. The main or only function seem to be to build a url?